src/testdrive/src/action/kafka/ingest.rs

```diff
@@ -130,145 +179,107 @@ impl Action for IngestAction {
+                // TODO: this would not work to send key-value input in bytes format
+                let mut tokens = Deserializer::from_str(&row).into_iter::<Value>();
+                let key_row = tokens.next();
+                let val_row = tokens.next();
 
-                        if tokens.next().is_some() || key_row.is_none() || val_row.is_none() {
-                            return Err(format!(
-                                "invalid row: {}; testdrive expects two json objects",
-                                row
-                            ));
-                        }
+                if tokens.next().is_some() || key_row.is_none() || val_row.is_none() {
+                    return Err(format!(
+                        "invalid row: {}; testdrive expects two json objects",
+                        row
+                    ));
+                }
 
-                        (Some(key_row.unwrap()), val_row.unwrap())
-                    } else {
-                        (None, serde_json::from_str(row))
-                    };
-                    let val = crate::format::avro::json_to_avro(
-                        &val_row.map_err(|e| format!("parsing avro datum: {}", e.to_string()))?,
-                        value_schema.top_node(),
-                    )?;
-                    // The first byte is a magic byte (0) that indicates the Confluent
-                    // serialization format version, and the next four bytes are a
-                    // 32-bit schema ID.
-                    //
-                    // https://docs.confluent.io/current/schema-registry/docs/serializer-formatter.html#wire-format
-                    val_buf.write_u8(0).unwrap();
-                    val_buf
-                        .write_i32::<NetworkEndian>(*value_schema_id)
-                        .unwrap();
-                    val_buf.extend(
-                        avro::to_avro_datum(&value_schema, val).map_err(|e| e.to_string())?,
-                    );
-                    if let Some((key_schema, key_schema_id)) = key_encode {
-                        let key = crate::format::avro::json_to_avro(
-                            &key_row
-                                .unwrap()
+                (
+                    Some(
+                        key_row
+                            .unwrap()
+                            .map_err(|e| format!("parsing avro datum: {}", e.to_string()))?
+                            .to_string(),
+                    ),
+                    val_row
+                        .unwrap()
+                        .map_err(|e| format!("parsing avro datum: {}", e.to_string()))?
+                        .to_string(),
+                )
+            } else {
+                (None, row.clone())
+            };
+
+            fn decode_into_buf(
+                encoder: &Encoder,
+                row: &str,
+                buf: &mut Vec<u8>,
+            ) -> Result<(), String> {
```

Comment:
This seems like it belongs as a method on `Encoder`. It would make this method much shorter, too! I'd recommend hoisting `Encoder` out of this function, and adding this as `Encoder::encode_into(&self, row: &str, buf: &mut Vec<u8>) -> Result<(), String>`.